src/Hive.Versioning/Parsing/ErrorMessages.cs

```diff
@@ -0,0 +1,883 @@
+﻿using System.Text;
+using Hive.Versioning.Resources;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Hive.Utilities;
+using static Hive.Versioning.StaticHelpers;
+
+#if !NETSTANDARD2_0
+using StringPart = System.ReadOnlySpan<char>;
+#else
+using StringPart = Hive.Utilities.StringView;
+#endif
+
+namespace Hive.Versioning.Parsing
+{
+    public static class ErrorMessages
+    {
+        // TODO: refactor to make data more easily passed up and down the callstack
+
+        private record GeneratedMessage(string Message,
+            string? Suggestion = null,
+            (long Start, long Length)? Span = null,
+            (long Start, long Length)? ApplyRange = null, // if ApplyRange is null, we use Span
+            bool ShowSuggestion = true);
+
+        private ref struct MessageInfo
+        {
+            public readonly StringPart Text;
+
+            public MessageInfo(in StringPart text)
+            {
+                Text = text;
+            }
+        }
+
+        private static (long Start, long Length) SpanFromReport<T>(in ActionErrorReport<T> report)
+            where T : struct
+            => (report.TextOffset, report.Length);
+
+        #region Version
+        public static string GetVersionErrorMessage(ref ParserErrorState<VersionParseAction> errors)
+        {
+            var reports = errors.ToArray();
+
+            var msgInfo = new MessageInfo(errors.InputText);
+            var msg = ProcessVersionErrorMessage(in msgInfo, reports);
+            return FormatMessage(in msgInfo, msg);
+        }
+
+        private static GeneratedMessage? ProcessVersionErrorMessage(in MessageInfo msgs, IReadOnlyList<ActionErrorReport<VersionParseAction>> reports)
+        {
+            var range = ScanForErrorRange(reports, VersionIsError);
+
+            if (range.Length == 0)
+                return null;
+
+            var ourTextStart = reports[0].TextOffset;
+
+            if (range.Length == 1)
+            {
+                var report = reports[range.Start];
+#pragma warning disable IDE0010 // Add missing cases
```

Comment:
Are these warnings worth being suppressed with justification instead? The repeated disable + justification is unfortunate, imo.

Reply:
You literally do it in the following method, curious why you chose to avoid it here.

Reply:
I believe that was done as I was incrementally implementing behaviour. Doing it this way meant that later switches would still get the warnings.

Reply:
Willing to change that though.

Reply:
Following up on that, I do feel like keeping them separate is better for methods which may contain multiple such switches.